naxsi_src/naxsi_runtime.c

```diff
@@ -2685,54 +2715,41 @@ ngx_http_naxsi_body_parse(ngx_http_request_ctx_t*     ctx,
       ngx_http_apply_rulematch_v_n(&nx_int__uncommon_url, ctx, r, NULL, NULL, BODY, 1, 0);
-      return;
-    }
   }
   /* 19 = echo -n "multipart/form-data" | wc -c */
-  else if (!ngx_strncasecmp(
-             r->headers_in.content_type->value.data, (u_char*)"multipart/form-data", 19)) {
+  else if (!ngx_strncasecmp(content_type_str, (u_char*)"multipart/form-data", 19)) {
     ngx_http_naxsi_multipart_parse(ctx, r, full_body, full_body_len);
   }
   /* 16 = echo -n "application/json" | wc -c */
-  else if (!ngx_strncasecmp(
-             r->headers_in.content_type->value.data, (u_char*)"application/json", 16)) {
+  else if (!ngx_strncasecmp(content_type_str, (u_char*)"application/json", 16)) {
     ngx_http_naxsi_json_parse(ctx, r, full_body, full_body_len);
   }
   /* 24 = echo -n "application/vnd.api+json" | wc -c */
-  else if (!ngx_strncasecmp(
-             r->headers_in.content_type->value.data, (u_char*)"application/vnd.api+json", 24)) {
+  else if (!ngx_strncasecmp(content_type_str, (u_char*)"application/vnd.api+json", 24)) {
     ngx_http_naxsi_json_parse(ctx, r, full_body, full_body_len);
   }
   /* 22 = echo -n "application/csp-report" | wc -c */
-  else if (!ngx_strncasecmp(
-             r->headers_in.content_type->value.data, (u_char*)"application/csp-report", 22)) {
+  else if (!ngx_strncasecmp(content_type_str, (u_char*)"application/csp-report", 22)) {
     ngx_http_naxsi_json_parse(ctx, r, full_body, full_body_len);
   } else {
     ngx_log_debug(NGX_LOG_DEBUG_HTTP, r->connection->log, 0, "[POST] Unknown content-type");
-    ngx_http_apply_rulematch_v_n(&nx_int__uncommon_content_type, ctx, r, NULL, NULL, BODY, 1, 0);
     /*
-    ** Only attempt to process "raw" body if id:nx_int__uncommon_content_type
-    *was
+    ** Only attempt to process "raw" body if id:nx_int__uncommon_content_type was
     ** whitelisted. Else, it should be blocking and stop processing here.
     */
-    if ((!ctx->block || ctx->learning) && !ctx->drop) {
-      ngx_log_debug(NGX_LOG_DEBUG_HTTP, r->connection->log, 0, "After uncommon content-type");
-      ngx_http_naxsi_rawbody_parse(ctx, r, full_body, full_body_len);
+    if (!ngx_http_apply_rulematch_v_n(
+          &nx_int__uncommon_content_type, ctx, r, NULL, NULL, BODY, 1, 0)) {
+      if ((!ctx->block || ctx->learning) && !ctx->drop) {
+        ngx_log_debug(NGX_LOG_DEBUG_HTTP, r->connection->log, 0, "After uncommon content-type");
+        ngx_http_naxsi_rawbody_parse(ctx, r, full_body, full_body_len);
+      }
     }
   }
+  ngx_pfree(r->pool, full_body);
 }
 
-/*
-** does : this is a 'main' function, all the stuff goes from here.
-**    to make it short, it does the following :
-** - if we got header rules, apply header_rules on each.
-** - apply generic_rules on url decoded URI.
-** - if we got get_rules and get args, apply get_rules varname/value couple.
-** - if we are in a POST/PUT request and we got body_rules, apply rules :)
-*/
 void
```

Comment:
why?

Reply:
Because ngx_http_naxsi_data_parse is main function, and ngx_http_naxsi_uri_parse is called from it
Most likely a copy-paste error